main.c

```diff
@@ -137,7 +142,68 @@ static void linux_boot(const struct boot_params boot_params)
     printf("  pref_address: 0x%llx\n", l_params->pref_address);
     printf("\n");
 
-    char *linux_header_end = (char *)(boot_params.kernel_addr + 0x201);
+    // Checks if a relocation of the kernel is required.
+    // We decided that the loader moves all kernels to their preferred address, even if they
+    // are relocatable, thus, could relocate themselves.
+    if (boot_params.kernel_addr == l_params->pref_address)
+    {
+        printf(
+            "The kernel is already loaded at its preferred address 0x%llx\n",
+            l_params->pref_address
+        );
+    } else {
+        printf(
+            "The kernel will be moved from 0x%lx to its preferred load address 0x%llx.\n",
+            boot_params.kernel_addr,
+            l_params->pref_address
+        );
+
+        uintptr_t kernel_begin_addr = l_params->pref_address;
+        uintptr_t kernel_end_addr = kernel_begin_addr + l_params->init_size;
+
+        // check if the space is valid in the memory map
+        uint8_t kernel_in_valid_mem_range = 0;
+        for (size_t i = 0; i < lib_sysinfo.n_memranges; i++) {
+            // the location must be in valid RAM
+            if (lib_sysinfo.memrange[i].type == CB_MEM_RAM) {
+                uintptr_t mem_range_begin = lib_sysinfo.memrange[i].base;
+                uintptr_t mem_range_end = mem_range_begin + lib_sysinfo.memrange[i].size;
+                if (kernel_begin_addr >= mem_range_begin
+                    && kernel_end_addr < mem_range_end) {
+                    kernel_in_valid_mem_range = 1;
+                }
+            }
+        }
+        die_on(
+            !kernel_in_valid_mem_range,
+            "the preferred load address is not in available RAM (according to memory map)"
+        );
+
+        // check that the kernel will not clash with the initrd after relocation
+        uintptr_t initrd_size = *(uint64_t *)boot_params.initrd_size_addr;
+        die_on(kernel_begin_addr >= boot_params.initrd_addr
+                   && l_params->pref_address < boot_params.initrd_addr + initrd_size,
+               "Kernel load address collides with initrd\n"
+        );
+
+        // check that the kernel will not clash with the cmdline after relocation
+        // value taken from: arch/x86/include/asm/setup.h:#define COMMAND_LINE_SIZE 2048
+        const size_t MAX_LEN = 2048;
+        uintptr_t cmdline_size = strnlen((char *)boot_params.cmdline_addr, MAX_LEN);
```

Comment:
Actually, since we already rely on protocol version 2.10 now, we can just take the `cmdline_size`, which is included since boot protocol version 2.06.